### ISSUE TITLE
Impossible to declare a session after a Python import

### ISSUE THREAD

**Issue by Julien281** [malmaud/TensorFlow.jl#403]
I would like to use the Python's [batch_norm function](https://www.tensorflow.org/api_docs/python/tf/layers/batch_normalization).

To do that, I can use the following commands to import a Python function:

```
using PyCall
@pyimport tensorflow as pythonTF
```
After that, I can't create a session of TensorFlow:

```
using TensorFlow
sess = Session(Graph())
```

returns the following error:

```
ERROR: could not load library "/home/jabou/.julia/v0.6/TensorFlow/src/../deps/usr/bin/libtensorflow"
/home/jabou/.julia/v0.6/TensorFlow/src/../deps/usr/bin/libtensorflow.so: undefined symbol: _ZNK6google8protobuf7Message11GetTypeNameB5cxx11Ev
```

I checked I can create a session without import a Python function, it works:

```
using PyCall
using TensorFlow
sess = Session(Graph())

2018-06-11 14:09:10.556001: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1030] Found device 0 with properties: 
name: TITAN Xp major: 6 minor: 1 memoryClockRate(GHz): 1.582
pciBusID: 0000:02:00.0
totalMemory: 11.91GiB freeMemory: 11.74GiB
2018-06-11 14:09:10.556053: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1120] Creating TensorFlow device (/device:GPU:0) -> (device: 0, name: TITAN Xp, pci bus id: 0000:02:00.0, compute capability: 6.1)
Session(Ptr{Void} @0x00007f41117ea450)

```

**Comment by oxinabox** [comment]
Yes, that is a thing I think.
I suspect because you can't load two different copies of libtensorflow into the same process,

@malmaud  has some medium-long  term plans for being able to use more of the functionality that the python client exposes.

Right now doing something like that is a bit of the wild-west.
**I'm not able to provide support for that kinda thing right now.
I have a thesis that is becoming very due.**

PyCalling tensorflow is possible, because we already have it isolated into its own process.


### Examples of calling python
A basic example of doing that is reading the version:
https://github.com/malmaud/TensorFlow.jl/blob/c253381a13319f5034d49669582348bb6428721f/src/core.jl#L63

However, things are more complicated when there is a `Graph` involved.
As you need to send the `Graph` to python, then run the operation, then import the `Graph` back.
The only example we have of that is the mechanics behind `gradients`

https://github.com/malmaud/TensorFlow.jl/blob/c253381a13319f5034d49669582348bb6428721f/src/core.jl#L1565-L1587

and

https://github.com/malmaud/TensorFlow.jl/blob/c253381a13319f5034d49669582348bb6428721f/src/py.jl#L74-L99


----------------------------
## Generalizing that to be able to call python

It wouldn't be impossible to abstract that into something a bit more general.
Untested code follows:




## Add to core.jl:
```
function call_python_function(py_function_name::Symbol, args..., kwargs...)
	arg_names = map(node_name, args)
	kwarg_names = map(node_name, values(kwargs))
	kwarg_keywords = collect(keys(kwargs))
    
    meta_graph = train.export_meta_graph()
    b = IOBuffer()
    writeproto(b, meta_graph)
    jl_graph_proto = @compat take!(b)
    node_protos, ret_names = fetch(@py_proc python_function($py_function_name, $jl_graph_proto, $arg_names, $kwarg_keywords, $kwarg_names))
    extend_graph(node_protos)
    out = []
    for name in ret_names
        if isa(name, String)
            push!(out, get_tensor_by_name(name))
        elseif isa(name, Void)
            push!(out, nothing)
        else
            push!(out, IndexedSlices(get_tensor_by_name(name[1]), get_tensor_by_name(name[2])+1))
        end
    end
	return out
end
```
### Add to py.jl
```
function python_function(py_function_name, jl_graph_proto, argnames, kwarg_keywords, kwarg_names)
    py_graph = make_py_graph(jl_graph_proto)

    to_py_node(node_name) = py_graph[:get_tensor_by_name](string(node_name[1], ":", node_name[2]-1))
    to_py_node(node_names::AbstractVector) = tuple(to_py_node.(node_names)...) #Need tuple as Vector will not be accepted
    to_py_node(::Void) = nothing

	args = map(to_py_node, args)
	kwarg_nodes = map(to_py_node, values(kwarg_names))
	kwargs = Dict(zip(kwarg_keywords, kwarg_nodes))
	
    @py_catch ret_nodes = py_tf[][py_function_name](args...; kwargs...)
    py_graph_def = py_graph[:as_graph_def]()
    ret_names = []
    for node in ret_nodes
        if node === nothing
            push!(ret_names, nothing)
            continue
        end
        try
            push!(ret_names, (node[:values][:name], node[:indices][:name]))
        catch
            push!(ret_names, node[:name])
        end
    end
    return to_protos(py_graph_def), ret_names
end
```




-----------------------------



It is really annoying that `batch_norm` is in Python, and not in libtensorflow.
Some of its underlying functionality  is in libtensorflow (`FusedBatchNormV2` iirc).
and I spent half a day trying to hook that up a few weeks back,
but I couldn't workout what glue it was needing
